### ISSUE TITLE
hw1

### ISSUE THREAD

**Issue by Oranga** [Oranga/ML2017FALL#1]
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time

start = time.time()
###
#input_file = 'C:/Users/Hao/Desktop/ML/hw1/train.csv'
#output_file = 'C:/Users/Hao/Desktop/ML/hw1/result.csv'
input_file = 'C:/Users/Oranga/Desktop/ML/hw1/train.csv'
output_file = 'C:/Users/Oranga/Desktop/ML/hw1/modelver8.csv'
###

x = np.zeros([12,20,15,18,9])
x_vld = np.zeros([10,18,9])
w = np.zeros([12,20,15,18,9])
y_head = np.zeros([12,20,15])
y_head_vld = np.zeros(10)
b = 0
iRand = 0
data = pd.read_csv(input_file, encoding='big5', index_col='測項')
x_total = list([data.loc[('AMB_TEMP')], data.loc[('CH4')], data.loc[('CO')], data.loc[('NMHC')], data.loc[('NO')], data.loc[('NO2')], data.loc[('NOx')], data.loc[('O3')], data.loc[('PM10')], data.loc[('PM2.5')], data.loc[('RAINFALL')].replace('NR',0), data.loc[('RH')], data.loc[('SO2')], data.loc[('THC')], data.loc[('WD_HR')], data.loc[('WIND_DIREC')], data.loc[('WIND_SPEED')], data.loc[('WS_HR')]])

randMon = np.random.permutation(12)
randDay = np.random.permutation(20)
randTime = np.random.permutation(15)

# 12 months
for i in range(12):
    # 20 days
    for j in range(20):
        # 15 data times
        for k in range(15):
            if (i in randMon[:10]) and (j in randDay[:10]) and (k in randTime[:10]) and (iRand<10):
                y_head_vld[iRand] = float(pd.to_numeric(data.loc[('PM2.5')][str(k+9)])[20*i+j:20*i+j+1])
                # 18 reference
                for h in range(18):
                    # 9 hours
                    for r in range(9):
                        x_vld[iRand] = float(pd.to_numeric(x_total[h][str(k+r)][20*i+j:20*i+j+1]))  
                iRand += 1
            else:
                y_head[i,j,k] = float(pd.to_numeric(data.loc[('PM2.5')][str(k+9)])[20*i+j:20*i+j+1])
                # 18 reference
                for h in range(18):
                    # 9 hours
                    for r in range(9):
                        x[i][j][k][h][r] = float(pd.to_numeric(x_total[h][str(k+r)][20*i+j:20*i+j+1]))
#%%
epoch = 10000
RT = 0
step = 1e-12
vldLoss = np.zeros(epoch)
loss = np.zeros(epoch)

while RT < epoch:
    y = b + np.sum(np.sum(w*x,axis = 3),axis = 3)
    y_vld = b + np.sum(np.sum(w[0,0,:10]*x_vld, axis = 1), axis = 1)
    
    loss[RT] = np.sqrt(np.sum((y_head-y)**2)/3590)
    vldLoss[RT] = np.sqrt(np.sum((y_head_vld-y_vld)**2)/10)
    
    slope_b = -2*np.sum(y_head-y)
    slope_w = slope_b*x
    b -= step*slope_b
    w -= step*slope_w     

    RT += 1           
#    print('%.2f' %lossAvg)
#    
plt.figure(1)
plt.subplot(211)
plt.plot(range(100,epoch), loss[100:])
plt.subplot(212)
plt.plot(range(100,epoch), vldLoss[100:])
plt.show()
end = time.time()
print('Running time = %.2f' %(end-start))

#%%
model = ['w/b',b]
for i in range(18):
    for j in range(9):
        model.append(w[0][0][0][i][j])                             
model = pd.DataFrame(model)
model.to_csv(output_file, index=False, header=False)
